specs/keccak256/variable-length-input.md

```diff
@@ -0,0 +1,122 @@
+# Keccak256 For Variable Length Input
+
+We describe the circuit that can process input of arbitrary size.
+
+## Plain Keccak behavior
+
+Keccak256 use a simple iterating sponge construction to handle a variable length input. The sponge absorbs 17 words(136 bytes) of inputs, applies the `Keccak-f` permutation, then absorb next 17 words. Once all parts of the inputs are absorbed, the sponge squeezes out the 32 bytes for the output.
+
+If the last part of the inputs is less than 136 bytes, a padding must be applied. The padding is a multi-rate padding, it pads single bit 1 followed by minimum number of bits 0 followed by a single bit 1, such that the result is 136 bytes.
+
+## Circuit behavior
+
+### Lookup Region
+
+This region splits the input to multiple parts, each part corresponds to their `Keccak-f` permutation round.
+This is also a lookup table for the other circuits to lookup the Keccak256 input to the output.
+
+Define:
+
+We use `Round` for the cells of the following columns in a row, it represents the information in a `keccak_f` round.
+
+The `curr: Round` and `next: Round` represent the current and the next row.
+
+Columns:
+
+- `hash_id` Sequential number for grouping different hash
+- `input_len` Length for correct padding
+- `input` 136 bytes to be absorbed in this round
+- `acc_len` How many length we have absorbed
+- `acc_input` Accumulatd bytes by random linear combination (in big-endian order)
+- `output` The base-2 `state[:4]` output from this round `keccak_f`
+- `is_end_result` This flag indicates the hash output is final. The purpose is for external circuit to lookup.
+
+Selector:
+
+- `is_last_round_of_circuit` Last round of the whole circuit (not hash), to avoid `next` wrap around.
+
+#### Checks
```

Comment:
I think we are missing a check for all of the rounds that will come after the padded one and which are essentially 0.

For all of these, we need to constrain the bytes to be 0 right? Or it does not affect since in the big circuit we will "dissable the permutation" when round > final_round_for_padding?

Reply:
I think I have an idea for 1 now. Instead of complicating the design to handle unused permutations, the prover should add a dummy hash job `hash_id=3` the use it for 0 inputs for 97 permutations.

Reply:
Yep, my concern was for 1. I'm not sure I follow you. Following @barryWhiteHat 's idea on that part, he purposed to do the following:

- If permutation contains initial input bytes -> `inp_len <= acc_len` for the last byte of the permutation:
1) ConditionallySelect(prev_output_state  || 0's) into actual state for the permutation computation
2) ConditionallySelect(actual_permutation_bytes || 0's) into actual permutation input
3) `final_result[0..N] += !is_last_permutation * out_state[0..N]
4) `running_sum_last_byte == output` NOTE that 3 and 4 are the same but with different formats.

I can elaborate it more. But will be something like saying, you conditionally copy to the next permutation as inputs `(state and next_inputs)` or a bunch of 0's. And you also conditionally add to `final_result` the result of the permutation depending on if it is the last one or not.

Point 2) sounds good as we discussed! :)